### PR TITLE
Allow subclasses to map exceptions to status codes.

### PR DESCRIPTION
**Pull request by garydgregory**
This tiny refactoring allow subclasses to map exceptions to status codes in the simplest manner.

### CODE REVIEW

**Comment by olamy** [comment]
@garydgregory you need to sign eclipse ECA first

**Comment by garydgregory** [comment]
I did. I see "You successfully submitted an Eclipse Contributor Agreement."

**Comment by olamy** [comment]
@garydgregory did you sign your commit?

**Comment by garydgregory** [comment]
No, I had no idea that was a requirement... can I sign it post facto?

**Comment by lachlan-roberts** [comment]
use `git commit --amend --signoff`

File: jetty-proxy/src/main/java/org/eclipse/jetty/proxy/AbstractProxyServlet.java [hunk 1]
```diff
@@ -639,16 +639,20 @@ protected void onProxyResponseFailure(HttpServletRequest clientRequest, HttpServ
...
             status = serverStatus;
         sendProxyResponseError(clientRequest, proxyResponse, status);
         
     }
 
+	protected int toStatus(Throwable failure) {
+		return failure instanceof TimeoutException ?
+            HttpStatus.GATEWAY_TIMEOUT_504 :
+                HttpStatus.BAD_GATEWAY_502;
+	}
```
Comment (sbordet) [inline jetty-proxy/src/main/java/org/eclipse/jetty/proxy/AbstractProxyServlet.java]: 1. rename to `proxyResponseStatus()`
2. braces on newline
3. align properly `return` and `:?`

**Comment by garydgregory** [comment]
I keep on updating mthe Eclipse ECA and it keeps staying "You successfully submitted an Eclipse Contributor Agreement." and I see a green checkmark in the STATUS pane next to "Eclipse Contributor Agreement" so I am not sure what more I need to do.

**Comment by sbordet** [comment]
@garydgregory please bear with the process, we need it to accept contributions.
All your commits are re-scanned and because you have one that was not compliant, the check will fail.

You need to squash all your commits into one and force-push this PR so that there will be only one commit.
This one commit must be signed-off and the email you signed it off with must be the same email you used for the Eclipse ECA.
I have verified that for `gardgregory@gmail.com` - which is what you signed your commits with - there is no ECA on record.
With what email did you sign the ECA?

**Comment by joakime** [comment]
There's 2 different email addresses you've submitted commits against in this PR ...

> Anonymous (ggregory@****.com) is not covered by the necessary legal agreements in order to proceed.
> Anonymous (gardgregory@****.com) is not covered by the necessary legal agreements in order to proceed.

Both of them are lacking a signed ECA.

**Comment by garydgregory** [comment]
When I go to the "Edit My Profile" page, the "Email Address" text field shows "garydgregory@gmail.com". What now?

**Comment by joakime** [comment]
> When I go to the "Edit My Profile" page, the "Email Address" text field shows "[garydgregory@gmail.com](mailto:garydgregory@gmail.com)". What now?

That is yet another email address. we now have 3 we are talking about.
The 2 you submitted the on the commits themselves. and this 3rd one in your Eclipse account.

If you can rebase (with squash) this entire PR then maybe we can get back to a single thing we need to address.

(I'm assuming your fork is checked out on the git remote called `origin`)
```
$ cd /path/to/my/jetty-9.4.x
$ git checkout jetty-9.4.x_proxy_with_custom_status_code
$ git rebase -i origin/jetty-9.4.x
# Double check the one and only commit in this branch, make sure it uses for BOTH the commit author and --sign-off the same email address `garydgregory@gmail.com`.
$ git push origin jetty-9.4.x_proxy_with_custom_status_code --force-with-lease
```

At the UI that pops up for `git rebase -i` make sure the first line says "pick", and all other line start with just "s" (squash).
And then follow the command line prompts to finish the rebase.

**Comment by garydgregory** [comment]
Thank you for your help. I squashed. There is now a single commit with garydgregory@gmail.com as the author and signed-off by marker.

**Comment by sbordet** [comment]
@garydgregory I did not check the author but GitHub reports your commit as if done by 2 different people.
So I guess the author is different from the signoff.
Your signoff is:
`Signed-off-by: Gary Gregory <gardgregory@gmail.com>`
which does not have a valid ECA.

**Comment by joakime** [comment]
Like @sbordet pointed out, the author of your commit and the signoff have 2 different email addresses (i suspect the signoff is a typo).

```
commit 392bb19a7df044a4d82784261b825f33101db9c2 (HEAD -> garydgregory-jetty-9.4.x_proxy_with_custom_status_code)
Author: Gary Gregory <garydgregory@gmail.com>
Date:   Sun May 5 19:12:54 2019 -0400

    Allow subclasses to map exceptions to status codes.
    
    Signed-off-by: Gary Gregory <gardgregory@gmail.com>
```

**Comment by garydgregory** [comment]
Dang, I'm lame! Fixed the signed-off. Thank you again for your patience.

**Comment by joakime** [comment]
Much better! the Eclipse ECA check is happy now. :+1: 
I'll leave it up to @sbordet to review the current state of this PR.

File: jetty-proxy/src/main/java/org/eclipse/jetty/proxy/AbstractProxyServlet.java [hunk 2]
```diff
@@ -568,13 +568,17 @@ protected void onClientRequestFailure(HttpServletRequest clientRequest, Request
...
         {
-            int status = failure instanceof TimeoutException ?
-                    HttpStatus.REQUEST_TIMEOUT_408 :
-                    HttpStatus.INTERNAL_SERVER_ERROR_500;
+            int status = clientRequestStatus(failure);
             sendProxyResponseError(clientRequest, proxyResponse, status);
         }
     }
 
+	protected int clientRequestStatus(Throwable failure) {
```
Comment (sbordet) [inline jetty-proxy/src/main/java/org/eclipse/jetty/proxy/AbstractProxyServlet.java]: Brace on newline.

File: jetty-proxy/src/main/java/org/eclipse/jetty/proxy/AbstractProxyServlet.java [hunk 3]
```diff
@@ -568,13 +568,17 @@ protected void onClientRequestFailure(HttpServletRequest clientRequest, Request
...
+		return failure instanceof TimeoutException ?
```
Comment (sbordet) [inline jetty-proxy/src/main/java/org/eclipse/jetty/proxy/AbstractProxyServlet.java]: Too much indentation for `return`. Use 4 spaces, no tabs.
Broken line gets indented 8 spaces.

File: jetty-proxy/src/main/java/org/eclipse/jetty/proxy/AbstractProxyServlet.java [hunk 4]
```diff
@@ -639,16 +643,21 @@ protected void onProxyResponseFailure(HttpServletRequest clientRequest, HttpServ
...
         int serverStatus = serverResponse == null ? status : serverResponse.getStatus();
         if (expects100Continue(clientRequest) && serverStatus >= HttpStatus.OK_200)
             status = serverStatus;
         sendProxyResponseError(clientRequest, proxyResponse, status);
         
     }
 
+	protected int proxyResponseStatus(Throwable failure) 
+	{
+		return failure instanceof TimeoutException ?
```
Comment (sbordet) [inline jetty-proxy/src/main/java/org/eclipse/jetty/proxy/AbstractProxyServlet.java]: Too much indentation for return. Use 4 spaces, no tabs.
Broken line gets indented 8 spaces.

**Comment by garydgregory** [comment]
I fixed the spacing locally, squahsed and force-pushed, and it still looks messed up on https://patch-diff.githubusercontent.com/raw/eclipse/jetty.project/pull/3625.diff

**Comment by sbordet** [comment]
@garydgregory please use spaces (not tabs) and look at the rest of the file for how the formatting should be.

**Comment by garydgregory** [comment]
I am going to close this PR and open a new one.

**Pull request closed without merging**